README.md
# *Herbie* lint for Rust [![Build Status][travis-svg]][travis] [![Crates.io][crate-svg]][crate] [![License][license-svg]][license]

## What

This plugin can add warnings or errors to your crate when using a numerically
unstable floating point expression.

Quick example of what you can get when compiling
[`tests/compile-fail/general/test.rs`][example]:
```rust
test.rs:40:5: 40:18 warning: Numerically unstable expression, #[warn(herbie)] on by default
test.rs:40     (a/b + c) * b;
               ^~~~~~~~~~~~~
test.rs:40:5: 40:18 help: Try this
test.rs:       (c * b) + a;
test.rs:67:5: 67:23 warning: Numerically unstable expression, #[warn(herbie)] on by default
test.rs:67     (a*a + b*b).sqrt();
               ^~~~~~~~~~~~~~~~~~
test.rs:67:5: 67:23 help: Try this
test.rs:       a.hypot(b);
test.rs:155:5: 155:30 warning: Numerically unstable expression, #[warn(herbie)] on by default
test.rs:155     (a+b).sin() - (a+b).cos();
                ^~~~~~~~~~~~~~~~~~~~~~~~~
test.rs:155:5: 155:30 help: Try this
test.rs:        (b.sin() * (a.sin() + a.cos())) - ((a.cos() - a.sin()) * b.cos());
```

As you can see, it will report numerically unstable expressions, and suggest a
(sometimes over-parenthesized) more stable correction.

## Usage
### Plugin
This is a `rustc` plugin, to use it, you need a *nightly* Rust.

You need a database of possible corrections for this plugin to work. The
database format is the same as [Herbie GHC Plugin (for Haskell)][ghc-herbie]
from which this plugin is inspired so [this file][ghc-herbie-db] should work.
Just put it in the same directory you call `cargo` or `rustc` from.

Add in your *Cargo.toml*:

```toml
[dependencies]
herbie-lint = "{{VERSION}}"
```

and in your crate:

```rust
#![feature(plugin)]
#![plugin(herbie_lint)]
```

See [*clippy*][clippy]'s [*Usage* section][clippy-usage] if you want to know
more and if you want more Rust lints.

### Configuration
If you don't want the plugin to lint a particular function or method, you can
mark it with the `#[herbie_ignore]` attribute:

```rust
fn foo(a: f64, b: f64, c: f64) -> f64 {
    (a/b + c) * b
    // This will suggest to use “(c * b) + a” instead.
}

#[herbie_ignore]
fn bar(a: f64, b: f64, c: f64) -> f64 {
    (a/b + c) * b
    // This won't.
}
```

You can also put a `Herbie.toml` file next to your `Cargo.toml` with the
following fields:
```toml
# Path to the database.
db_path = "Herbie.db"

# The seed use by Herbie. If not provided, a fixed seed will be used. Fixing
# the seed ensures deterministic builds.
herbie_seed = "#(1461197085 2376054483 1553562171 1611329376 2497620867 2308122621)"

# Allow the plugin to call Herbie on unknown expressions. Positive results from
# Herbie will be cached in the database. WARNING: Herbie is slow.
# If ‘true’, the plugin will fail if it cannot find the executable.
# If ‘false’, the plugin will not try to run Herbie.
# By default, the plugin will call the executable only if it's found, but won't
# complain otherwise.
use_herbie = false

# Maximum time in seconds that Herbie is allowed to play with an expression. If
# null, allow Herbie to run indefinitely. Default is two minutes.
timeout = 120
```

More information about calling Herbie can be found in the
[wiki][wiki-herbie-inout].

## Acknowledgment
Thanks to @llogiq for [the idea][idea].

[clippy-usage]: https://github.com/Manishearth/rust-clippy#usage
[clippy]: https://github.com/Manishearth/rust-clippy
[crate-svg]: https://img.shields.io/crates/v/herbie-lint.svg
[crate]: https://crates.io/crates/herbie-lint/
[example]: https://github.com/mcarton/rust-herbie-lint/blob/master/tests/compile-fail/general/test.rs
[ghc-herbie-db]: https://github.com/mikeizbicki/HerbiePlugin/blob/master/data/Herbie.db?raw=true
[ghc-herbie]: https://github.com/mikeizbicki/HerbiePlugin
[idea]: https://github.com/Manishearth/rust-clippy/issues/346
[license-svg]: https://img.shields.io/crates/l/herbie-lint.svg
[license]: https://github.com/mcarton/rust-herbie-lint/blob/master/LICENSE
[travis-svg]: https://travis-ci.org/mcarton/rust-herbie-lint.svg
[travis]: https://travis-ci.org/mcarton/rust-herbie-lint/
[wiki-herbie-inout]: https://github.com/mcarton/rust-herbie-lint/wiki#how-to-use-herbie-optional


CHANGELOG.md
# Change Log

## 1.0.7 — 2016-04-08
* Rustup to *rustc 1.9.0-nightly (7979dd608 2016-04-07)*

## 1.0.6 — 2016-03-27
* Rustup to *rustc 1.9.0-nightly (d5a91e695 2016-03-26)*

## 1.0.5 — 2015-02-12
* Rustup to *rustc 1.8.0-nightly (4b2c7030f 2016-02-12)*

## 1.0.4 — 2015-01-26
* Implement a timeout when executing Herbie.
* More documentation and wiki.

## 1.0.3 — 2015-01-25
* Fix bug in Rust expression reconstruction which displayed Herbie function
  names.

## 1.0.2 — 2015-01-09
* Add the `herbie_ignore` attribute.

## 1.0.1 — 2015-01-09
* Support for `Herbie.toml` improved.
* Support for calling `herbie-inout` if available.
* Better error handling.


tests/lisp.rs
#![feature(box_syntax, rustc_private)]

#[allow(plugin_as_library)]
extern crate herbie_lint;
extern crate rustc;

use herbie_lint::lisp::{LispExpr, Parser, ParseError};
use herbie_lint::lisp::LispExpr::*;
use rustc::hir::BinOp_::*;
use rustc::hir::UnOp::*;

pub fn parse(s: &str) -> Result<LispExpr, ParseError> {
    let mut parser = Parser::new();
    parser.parse(s)
}

#[test]
fn test_parser() {
    assert_eq!(parse(""), Err(ParseError::EOE));

    assert_eq!(parse("0."), Ok(Lit(0.)));
    assert_eq!(parse("herbie"), Ok(Ident(0)));
    assert_eq!(parse("(+ 0. herbie1)"), Ok(Binary(BiAdd, box Lit(0.), box Ident(0))));

    let valid = &[
        ("0", Lit(0.)),
        ("herbie0", Ident(0)),
        ("(+ herbie0 herbie1)", Binary(BiAdd, box Ident(0), box Ident(1))),
        ("(- herbie0)", Unary(UnNeg, box Ident(0))),
        ("(- herbie0 herbie1)", Binary(BiSub, box Ident(0), box Ident(1))),
        ("(cos 1)", Fun("cos".into(), vec![Lit(1.)])),
        ("(cos herbie0)", Fun("cos".into(), vec![Ident(0)])),
        ("(log1p (cos herbie0))", Fun("log1p".into(), vec![Fun("cos".into(), vec![Ident(0)])])),
    ];

    for &(s, ref e) in valid {
        assert_eq!(parse(s).as_ref(), Ok(e));
        assert_eq!(s, e.to_lisp("herbie"));
    }


    assert_eq!(parse("(+ 0. 0.) foobar"), Err(ParseError::EOE));
    assert_eq!(parse("("), Err(ParseError::EOE));
    assert_eq!(parse("0.eee"), Err(ParseError::Float));


    assert_eq!(parse("+"), Err(ParseError::Unexpected('+')));
    assert_eq!(parse("(+)"), Err(ParseError::Unexpected(')')));
    assert_eq!(parse("(+ 0.)"), Err(ParseError::Arity));
    assert_eq!(parse("(+ 0. 0."), Err(ParseError::Expected(')')));
    assert_eq!(parse("(+ 0. 0. 0.)"), Err(ParseError::Expected(')')));
    assert_eq!(parse("(+ 0. 0. 0."), Err(ParseError::Expected(')')));

    assert_eq!(parse("-"), Err(ParseError::Unexpected('-')));
    assert_eq!(parse("(-)"), Err(ParseError::Unexpected(')')));
    assert_eq!(parse("(- 0. 0. 0.)"), Err(ParseError::Expected(')')));


    assert_eq!(parse("(cos)"), Err(ParseError::Arity));
    assert_eq!(parse("(cos"), Err(ParseError::Expected(')')));
    assert_eq!(parse("(cos 1. 2.)"), Err(ParseError::Arity));
    assert_eq!(parse("(cos 1. 2. 3."), Err(ParseError::Expected(')')));

    assert_eq!(parse("(foo 1.)"), Err(ParseError::Ident));
    assert_eq!(parse("(foocos 1.)"), Err(ParseError::Ident));
    assert_eq!(parse("(cosfoocos 1.)"), Err(ParseError::Ident));

    let mut parser = Parser::new();
    assert_eq!(parser.parse("(* (+ (/ herbie0 herbie1) herbie2) herbie1)"), Ok(
        Binary(BiMul,
            box Binary(BiAdd,
                box Binary(BiDiv, box Ident(0), box Ident(1)),
                box Ident(2)
            ),
            box Ident(1)
        )
    ));

    assert_eq!(parser.parse("(+ (* herbie2 herbie1) herbie0)"), Ok(
        Binary(BiAdd,
            box Binary(BiMul, box Ident(2), box Ident(1)),
            box Ident(0)
        )
    ));

    assert_eq!(parse("(+ (* herbie2 herbie1) herbie0)"), Ok(
        Binary(BiAdd,
            box Binary(BiMul, box Ident(0), box Ident(1)),
            box Ident(2)
        )
    ));
}


tests/test.rs
extern crate compiletest_rs as compiletest;

use std::env::{current_dir, set_var, var};
use std::fs::{copy, read_dir, remove_file};
use std::path::{Path, PathBuf};
use std::process::{Command, Output};

fn run_mode(mode: &'static str, dir: PathBuf, target_dir: &Path) {
    let mut config = compiletest::default_config();

    config.target_rustcflags = Some(format!("-L {}", target_dir.to_str().unwrap()));

    if let Ok(name) = var::<&str>("TESTNAME") {
        config.filter = Some(name.to_owned())
    }

    let cfg_mode = mode.parse().ok().expect("Invalid mode");
    config.mode = cfg_mode;

    config.src_base = current_dir().unwrap();

    let db_orig = Path::new("Herbie.orig.db");
    let has_db_orig = db_orig.exists();

    if has_db_orig {
        copy(db_orig, "Herbie.db").is_ok();
    }

    // Test dirs might have its own "herbie-inout" script
    set_var("PATH", format!("{}:{}", var("PATH").unwrap(), dir.to_str().unwrap()));

    compiletest::run_tests(&config);

    if has_db_orig {
        let Output { status: status_dest, stdout: stdout_dest, stderr: stderr_dest } =
            Command::new("sqlite3")
            .arg("Herbie.dest.db").arg("select * from HerbieResults;")
            .output()
            .unwrap()
        ;

        let Output { status, stdout, stderr } =
            Command::new("sqlite3")
            .arg("Herbie.db").arg("select * from HerbieResults;")
            .output()
            .unwrap()
        ;

        assert_eq!(stdout_dest, stdout);

        assert!(stderr.is_empty(), "{}", String::from_utf8_lossy(&stderr));
        assert!(stderr_dest.is_empty(), "{}", String::from_utf8_lossy(&stderr));

        assert!(status_dest.success());
        assert!(status.success());

        remove_file("Herbie.db").unwrap();
    }
}

#[test]
fn compile_test() {
    let cwd = current_dir().unwrap();
    let target_dir = cwd.join("target/debug/");

    for dir in read_dir("tests/compile-fail").unwrap() {
        let dir = cwd.join(dir.unwrap().path());
        run_mode("compile-fail", dir, &target_dir);
    }
}


tests/compile-fail/test.rs
#![feature(plugin)]
#![plugin(herbie_lint)]

#![deny(herbie)]

struct Foo { a: f64, b: f64 }

fn get_f64() -> f64 {
    4.2
}

fn get_tup() -> (f64, f64) {
    (0., 0.)
}

fn get_struct() -> Foo {
    Foo { a: 0., b: 0. }
}

fn integers() {
    1;
    (1/2 + 3) * 2;
}

fn main() {
    integers();

    let (a, b, c) = (0., 0., 0.);

    a*a + b*b;

    (a/a + a) * a;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (a * a) + a
    (a/b + a) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (a * b) + a
    (a/b + c) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (c * b) + a

    (a/b + c) * a;

    (0./1. + 2.) * 1.;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (2. * 1.) + 0.
    (1./1. + 2.) * 1.;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (2. * 1.) + 1.
    (1./1. + 1.) * 1.;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (1. * 1.) + 1.

    (0./1. + a) * 1.;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (a * 1.) + 0.
    (0./a + 2.) * a;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (2. * a) + 0.

    (a/b + get_f64()) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION ((get_f64()) * b) + a
    (a/b + (4.5f64).sqrt()) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (((4.5f64).sqrt()) * b) + a
    (a/b + (42 as f64)) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION ((42 as f64) * b) + a
    (a/b + { 42. }) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (({ 42. }) * b) + a

    (a/get_f64() + c) * get_f64();
    (a/{ 42. } + c) * { 42. };

    (4.5f64).abs();

    //(a*a + b*b).sqrt();
    //**^ ERROR
    //**| HELP Try this
    //**| SUGGESTION a.hypot(b)

    //(a*a + (-4. * -4.)).sqrt();
    //**^ ERROR
    //**| HELP Try this
    //**| SUGGESTION a.hypot(-4.))

    a.floor();

    (a/b + c.floor()) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION ((c.floor()) * b) + a

    let d = (0., 0.);

    (a/b + d.0) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (d.0 * b) + a

    (a/b + get_tup().0) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION ((get_tup().0) * b) + a

    (a/d.0 + c) * d.0;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (c * d.0) + a

    (a/d.0 + c) * d.1;
    (a/get_tup().0 + c) * get_tup().0;

    let e = get_struct();

    (a/b + e.a) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (e.a * b) + a

    (a/b + get_struct().a) * b;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION ((get_struct().a) * b) + a

    (a/e.a + c) * e.a;
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (c * e.a) + a

    (a/get_struct().a + c) * get_struct().a;
    (a/e.a + c) * e.b;

    1. - a.cos();
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (a.sin() * a.sin()) / a.cos().log1p().exp()

    (a+b).sin() - (a+b).cos();
    //~^ ERROR
    //~| HELP Try this
    //~| SUGGESTION (b.sin() * (a.sin() + a.cos())) - ((a.cos() - a.sin()) * b.cos())
}


src/lib.rs
#![feature(box_syntax)]
#![feature(plugin_registrar)]
#![feature(rustc_private)]

#![cfg_attr(feature="clippy", feature(plugin))]
#![cfg_attr(feature="clippy", plugin(clippy))]

#![deny(
    missing_copy_implementations,
    missing_debug_implementations,
    trivial_casts,
    trivial_numeric_casts,
    unsafe_code,
    unused_import_braces,
    unused_qualifications,
)]

extern crate itertools;
extern crate rusqlite;
#[macro_use]
extern crate rustc;
#[macro_use]
extern crate rustc_plugin;
extern crate rustc_serialize;
extern crate syntax;
extern crate toml;
extern crate wait_timeout;

use syntax::feature_gate::AttributeType;
use rustc_plugin::Registry;

mod conf;
pub mod lint;
pub mod lisp;
mod utils;

#[plugin_registrar]
pub fn plugin_registrar(reg: &mut Registry) {
    reg.register_late_lint_pass(box lint::Herbie::new());

    reg.register_attribute("herbie_ignore".into(), AttributeType::Whitelisted);
}


src/lint.rs
use conf;
use itertools::Itertools;
use lisp::LispExpr;
use lisp;
use rusqlite as sql;
use rustc::hir::*;
use rustc::hir::map::Node;
use rustc::lint::{LateContext, LintArray, LintContext, LintPass, LateLintPass};
use rustc::ty::TypeVariants;
use std::borrow::Cow;
use std::io::{Read, Write};
use std::process::{Command, Stdio};
use std::time::Duration;
use std;
use syntax::ast::MetaItemKind;
use syntax::ast::{Attribute, FloatTy};
use wait_timeout::ChildExt;

#[derive(Debug, Default)]
pub struct Herbie {
    conf: Option<conf::Conf>,
    initialized: bool,
    subs: Vec<(LispExpr, LispExpr)>,
}

#[derive(Debug)]
pub enum InitError {
    Conf {
        error: conf::ConfError,
    },
    SQL {
        error: sql::Error,
    },
}

impl From<conf::ConfError> for InitError {
    fn from(err: conf::ConfError) -> InitError {
        InitError::Conf { error: err }
    }
}

impl From<sql::Error> for InitError {
    fn from(err: sql::Error) -> InitError {
        InitError::SQL { error: err }
    }
}

impl std::fmt::Display for InitError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match *self {
            InitError::Conf { ref error } => write!(f, "Configuration error: {}", error),
            InitError::SQL { ref error } => write!(f, "Got SQL error: {}", error),
        }
    }
}

impl Herbie {
    pub fn new() -> Herbie {
        Herbie::default()
    }

    pub fn init(&mut self) -> Result<(), InitError> {
        if self.initialized {
            return Ok(())
        }

        self.initialized = true;

        let conf = try!(conf::read_conf());
        let connection = try!(sql::Connection::open_with_flags(
            conf.db_path.as_ref(), sql::SQLITE_OPEN_READ_ONLY
        ));
        let mut query = try!(connection.prepare("SELECT * FROM HerbieResults"));

        self.subs = try!(query.query(&[])).filter_map(|row| {
            match row {
                Ok(row) => {
                    let cmdin : String = row.get(1);
                    let cmdout : String = row.get(2);
                    // row.get(3) is opts ↔ Herbies options
                    let errin = row.get_checked(4).unwrap_or(0.);
                    let errout = row.get_checked(5).unwrap_or(0.);

                    if cmdin == cmdout || errin <= errout {
                        return None;
                    }

                    let mut parser = lisp::Parser::new();
                    match parser.parse(&cmdin) {
                        Ok(cmdin) => {
                            match parser.parse(&cmdout) {
                                Ok(cmdout) => {
                                    Some((cmdin, cmdout))
                                }
                                Err(..) => None,
                            }
                        }
                        Err(..) => None,
                    }
                }
                Err(..) => None,
            }
        }).collect();

        self.conf = Some(conf);

        Ok(())
    }

}

declare_lint!(pub HERBIE, Warn,
              "checks for numerical instability");

impl LintPass for Herbie {
    fn get_lints(&self) -> LintArray {
        lint_array!(HERBIE)
    }
}

impl LateLintPass for Herbie {
    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
        fn is_herbie_ignore(attr: &Attribute) -> bool {
            if let MetaItemKind::Word(ref word) = attr.node.value.node {
                word == &"herbie_ignore"
            }
            else {
                false
            }
        }

        let attrs = match cx.tcx.map.find(cx.tcx.map.get_parent(expr.id)) {
            Some(Node::NodeItem(item)) => &item.attrs,
            Some(Node::NodeTraitItem(item)) => &item.attrs,
            Some(Node::NodeImplItem(item)) => &item.attrs,
            _ => panic!("In herbie-lint: how did I get there?"),
        };

        if attrs.iter().any(is_herbie_ignore) {
            return;
        }

        let ty = cx.tcx.expr_ty(expr);

        if ty.sty != TypeVariants::TyFloat(FloatTy::F64) {
            return;
        }

        if let Err(err) =  self.init() {
            cx.span_lint_note(
                HERBIE,
                cx.krate.span,
                "Could not initialize Herbie-Lint",
                cx.krate.span,
                &err.to_string()
            );
            return;
        }

        let mut got_match = false;
        for &(ref cmdin, ref cmdout) in &self.subs {
            if let Some(bindings) = LispExpr::match_expr(expr, cmdin) {
                report(cx, expr, cmdout, &bindings);
                got_match = true;
            }
        }

        let conf = self.conf.as_ref().expect("Configuration should be read by now");
        if !got_match && conf.use_herbie != conf::UseHerbieConf::No {
            if let Err(err) = try_with_herbie(cx, expr, &conf) {
                cx.span_lint(HERBIE, expr.span, &err);
            }
        }
    }
}

fn try_with_herbie(cx: &LateContext, expr: &Expr, conf: &conf::Conf) -> Result<(), Cow<'static, str>> {
    let (lisp_expr, nb_ids, bindings) = match LispExpr::from_expr(expr) {
        Some(r) => r,
        // not an error, the expression might for example contain a function unknown to Herbie
        None => return Ok(()),
    };

    if lisp_expr.depth() <= 2 {
        return Ok(());
    }

    let seed: &str = &conf.herbie_seed;
    let mut command = Command::new("herbie-inout");
    let command = command
        .arg("--seed").arg(seed)
        .arg("-o").arg("rules:numerics")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
    ;

    let mut child = match command.spawn() {
        Ok(child) => child,
        Err(err) => {
            return if conf.use_herbie == conf::UseHerbieConf::Yes {
                // TODO: wiki
                Err(format!("Could not call Herbie: {}", err).into())
            }
            else {
                Ok(())
            };
        }
    };

    // TODO: link to wiki about Herbie.toml
    cx.sess().diagnostic().span_note_without_error(
        expr.span,
        "Calling Herbie on the following expression, it might take a while"
    );

    let params = (0..nb_ids).map(|id| format!("herbie{}", id)).join(" ");
    let cmdin = lisp_expr.to_lisp("herbie");
    let lisp_expr = format!("(lambda ({}) {})\n", params, cmdin);
    let lisp_expr = lisp_expr.as_bytes();
    child.stdin
        .as_mut().expect("Herbie-inout's stdin not captured")
        .write(lisp_expr).expect("Could not write on herbie-inout's stdin")
    ;

    match conf.timeout {
        Some(timeout) => {
            match child.wait_timeout(Duration::from_secs(timeout as u64)) {
                Ok(Some(status)) if status.success() => (),
                Ok(Some(status)) => {
                    return Err(format!("herbie-inout did not return successfully: status={}", status).into());
                }
                Ok(None) => {
                    cx.sess().diagnostic().span_note_without_error(expr.span, "Herbie timed out");
                    return Ok(());
                }
                Err(err) => {
                    return Err(format!("herbie-inout did not return successfully: {}", err).into());
                }
            }
        }
        None => {
            match child.wait() {
                Ok(status) if status.success() => (),
                Ok(status) => {
                    return Err(format!("herbie-inout did not return successfully: status={}", status).into());
                }
                Err(err) => {
                    return Err(format!("herbie-inout did not return successfully: {}", err).into());
                }
            }
        }
    }

    let mut stdout = if let Some(output) = child.stdout {
        output
    }
    else {
        return Err("cannot capture herbie-inout output".into());
    };

    let mut output = String::new();
    if let Err(err) = stdout.read_to_string(&mut output) {
        return Err(format!("cannot read output: {}", err).into());
    }

    let mut output = output.lines();

    let parse_error = |s: Option<&str>| -> Option<f64> {
        match s {
            Some(s) => {
                match s.split(' ').last().map(str::parse::<f64>) {
                    Some(Ok(f)) => Some(f),
                    _ => None,
                }
            }
            _ => None,
        }
    };

    let (errin, errout, cmdout) = match (parse_error(output.next()), parse_error(output.next()), output.next()) {
        (Some(errin), Some(errout), Some(cmdout)) => {
            (errin, errout, cmdout)
        }
        _ => {
            return Err("Could not parse herbie-inout output".into())
        }
    };


    if errin <= errout {
        return Ok(());
    }

    let mut parser = lisp::Parser::new();
    let cmdout = match parser.parse(cmdout) {
        Ok(cmdout) => cmdout,
        _ => return Err("Could not understand herbie-inout cmdout".into()),
    };

    report(cx, expr, &cmdout, &bindings);
    save(conf, &cmdin, &cmdout, "", errin, errout)
        .map_err(|err| format!("Could not save database, got SQL error {}", err).into())
}

fn report(cx: &LateContext, expr: &Expr, cmdout: &LispExpr, bindings: &lisp::MatchBindings) {
    cx.struct_span_lint(HERBIE, expr.span, "Numerically unstable expression")
      .span_suggestion(expr.span, "Try this", cmdout.to_rust(cx, &bindings))
      .emit();
}

fn save(
    conf: &conf::Conf,
    cmdin: &str, cmdout: &LispExpr,
    seed: &str,
    errin: f64, errout: f64
) -> Result<(), sql::Error> {
    let connection = try!(sql::Connection::open_with_flags(
        conf.db_path.as_ref(), sql::SQLITE_OPEN_READ_WRITE
    ));

    try!(connection.execute("INSERT INTO HerbieResults (cmdin, cmdout, opts, errin, errout)
                             VALUES ($1, $2, $3, $4, $5)",
                            &[&cmdin, &cmdout.to_lisp("herbie"), &seed, &errin, &errout]));

    Ok(())
}


src/conf.rs
use std;
use std::borrow::Cow;
use std::fs::File;
use std::io::{Error as IOError, Read};
use toml;

// That's the default seed used by the Haskell plugin
const DEFAULT_HERBIE_SEED: &'static str = "#(1461197085 2376054483 1553562171 1611329376 \
                                           2497620867 2308122621)";
const DEFAULT_DB_PATH: &'static str = "Herbie.db";
const DEFAULT_TIMEOUT: u32 = 120;

#[derive(Debug, RustcDecodable)]
pub struct UxConf {
    /// Path to the database. Defaults to "Herbie.db".
    pub db_path: Option<String>,
    /// The seed use by Herbie. If not provided, a fixed seed will be used. Fixing the seed ensures
    /// deterministic builds.
    pub herbie_seed: Option<String>,
    /// Maximum time in seconds that Herbie is allowed to play with an expression. If null, allow
    /// Herbie to run indefinitely. Default is two minutes.
    pub timeout: Option<u32>,
    /// Allow the plugin to call Herbie on unknown expressions. Positive results from Herbie will
    /// be cached in the database.
    /// If ‘true’, the plugin will fail if it cannot find the executable.
    /// If ‘false’, the plugin will not try to run Herbie.
    /// By default, the plugin will call the executable only if it's found, but won't complain
    /// otherwise.
    pub use_herbie: Option<bool>,
}

#[derive(Debug, Eq, PartialEq)]
pub enum UseHerbieConf {
    Default,
    No,
    Yes,
}

#[derive(Debug)]
pub struct Conf {
    pub db_path: Cow<'static, str>,
    pub herbie_seed: Cow<'static, str>,
    pub timeout: Option<u32>,
    pub use_herbie: UseHerbieConf,
}

impl Default for Conf {
    fn default() -> Conf {
        Conf {
            db_path: DEFAULT_DB_PATH.into(),
            herbie_seed: DEFAULT_HERBIE_SEED.into(),
            timeout: Some(DEFAULT_TIMEOUT),
            use_herbie: UseHerbieConf::Default,
        }
    }
}

impl From<UxConf> for Conf {
    fn from(ux: UxConf) -> Conf {
        Conf {
            db_path: ux.db_path.map_or(DEFAULT_DB_PATH.into(), Into::into),
            herbie_seed: ux.herbie_seed.map_or(DEFAULT_HERBIE_SEED.into(), Into::into),
            timeout: ux.timeout.map_or(Some(DEFAULT_TIMEOUT), |t| {
                if t == 0 {
                    None
                }
                else {
                    Some(t)
                }
            }),
            use_herbie: ux.use_herbie.map_or(UseHerbieConf::Default, |u| {
                if u {
                    UseHerbieConf::Yes
                }
                else {
                    UseHerbieConf::No
                }
            }),
        }
    }
}

#[derive(Debug)]
pub enum ConfError {
    Io {
        error: IOError,
    },
    Parse,
}

impl std::fmt::Display for ConfError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match *self {
            ConfError::Io{ ref error } => write!(f, "Error reading Herbie.toml: {}", error),
            ConfError::Parse => write!(f, "Syntax error in Herbie.toml"),
        }
    }
}

impl From<IOError> for ConfError {
    fn from(err: IOError) -> ConfError {
        ConfError::Io { error: err }
    }
}

pub fn read_conf() -> Result<Conf, ConfError> {
    if let Ok(mut conf) = File::open("Herbie.toml") {
        let mut buffer = String::new();
        try!(conf.read_to_string(&mut buffer));

        if let Some(conf) = toml::decode_str::<UxConf>(&buffer) {
            Ok(conf.into())
        }
        else {
            Err(ConfError::Parse)
        }
    }
    else {
        Ok(Conf::default())
    }
}
